src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSpatialDatatype.java

```diff
@@ -184,23 +203,544 @@ protected void serializeToWkb(boolean excludeZMFromWKB, SQLServerSpatialDatatype
+                     * and SEGMENT_LINEs only add up additional points.
+                     * 
+                     */
+                    if (segments[currentWKBSegmentIndex].getSegmentType() == SEGMENT_FIRST_ARC) {
+                        int numberOfPointsInStructure = 3;
+                        currentWKBSegmentIndex++;
+                        while (currentWKBSegmentIndex < segments.length
+                                && segments[currentWKBSegmentIndex].getSegmentType() != SEGMENT_FIRST_ARC
+                                && segments[currentWKBSegmentIndex].getSegmentType() != SEGMENT_FIRST_LINE) {
+                            numberOfPointsInStructure = numberOfPointsInStructure + 2;
+                            currentWKBSegmentIndex++;
+                        }
+                        buf.put(endian);
+                        buf.putInt(InternalSpatialDatatype.CIRCULARSTRING.getTypeCode());
+                        buf.putInt(numberOfPointsInStructure);
+                        if (originalRemainingStructureCount != remainingStructureCount) {
+                            currentWKBPointIndex--;
+                        }
+                        addCoordinateToBuffer(buf, numberOfPointsInStructure);
+                    } else if (segments[currentWKBSegmentIndex].getSegmentType() == SEGMENT_FIRST_LINE) {
+                        int numberOfPointsInStructure = 2;
+                        currentWKBSegmentIndex++;
+                        while (currentWKBSegmentIndex < segments.length
+                                && segments[currentWKBSegmentIndex].getSegmentType() != SEGMENT_FIRST_ARC
+                                && segments[currentWKBSegmentIndex].getSegmentType() != SEGMENT_FIRST_LINE) {
+                            numberOfPointsInStructure++;
+                            currentWKBSegmentIndex++;
+                        }
+                        buf.put(endian);
+                        buf.putInt(InternalSpatialDatatype.LINESTRING.getTypeCode());
+                        buf.putInt(numberOfPointsInStructure);
+                        if (originalRemainingStructureCount != remainingStructureCount) {
+                            currentWKBPointIndex--;
+                        }
+                        addCoordinateToBuffer(buf, numberOfPointsInStructure);
+                    } else {
+                        // CompoundCurves should start with first arc or first line, it should not come here
+                        break;
+                    }
+                    break;
+                default:
+                    // Shouldn't be possible to come here, existing Geometry/Geography doesn't support other cases
+                    break;
+            }
+            remainingStructureCount--;
+        }
+    }
+
+    private int calculateNumPointsInThisFigure() {
+        if (figures.length == 0) {
+            return 0;
+        }
+
+        return (currentWKBFigureIndex == figures.length - 1)
```

Comment:
Can you add comment explaining what this block does?

Reply:
Absolutely.